packages/web-client/app/resources/safes.ts

```diff
@@ -1,34 +1,291 @@
+abstract class TrackedSafe {
+  toJSON() {
+    throw new Error('Tracked safes do not support conversion to JSON');
+  }
+}
+
+// https://stackoverflow.com/questions/57324324/how-to-prevent-requiredt-in-typescript-from-removing-undefined-from-the-ty
+type RequiredKeepUndefined<T> = { [K in keyof T]-?: [T[K]] } extends infer U
+  ? U extends Record<keyof U, [any]>
+    ? { [K in keyof U]: U[K][0] }
+    : never
+  : never;
+
+export class TrackedPrepaidCardSafe
+  extends TrackedSafe
+  implements RequiredKeepUndefined<PrepaidCardSafe>
+{
+  type = 'prepaid-card' as PrepaidCardSafe['type'];
+  @tracked address!: PrepaidCardSafe['address'];
+  @tracked createdAt!: PrepaidCardSafe['createdAt'];
+  @tracked tokens!: PrepaidCardSafe['tokens'];
+  @tracked owners!: PrepaidCardSafe['owners'];
+  @tracked issuingToken!: PrepaidCardSafe['issuingToken'];
+  @tracked spendFaceValue!: PrepaidCardSafe['spendFaceValue'];
+  @tracked prepaidCardOwner!: PrepaidCardSafe['prepaidCardOwner'];
+  @tracked hasBeenUsed!: PrepaidCardSafe['hasBeenUsed'];
+  @tracked issuer!: PrepaidCardSafe['issuer'];
+  @tracked reloadable!: PrepaidCardSafe['reloadable'];
+  @tracked transferrable!: PrepaidCardSafe['transferrable'];
+  @tracked customizationDID: PrepaidCardSafe['customizationDID'];
+
+  constructor(safe: PrepaidCardSafe) {
+    super();
+    this.update(safe);
+  }
+
+  update(safe: PrepaidCardSafe) {
+    this.address = safe.address;
+    this.createdAt = safe.createdAt;
+    this.tokens = safe.tokens;
+    this.owners = safe.owners;
+    this.issuingToken = safe.issuingToken;
+    this.spendFaceValue = safe.spendFaceValue;
+    this.prepaidCardOwner = safe.prepaidCardOwner;
+    this.hasBeenUsed = safe.hasBeenUsed;
+    this.issuer = safe.issuer;
+    this.reloadable = safe.reloadable;
+    this.transferrable = safe.transferrable;
+    this.customizationDID = safe.customizationDID;
+  }
+}
+
+export class TrackedMerchantSafe
+  extends TrackedSafe
+  implements RequiredKeepUndefined<MerchantSafe>
+{
+  type = 'merchant' as MerchantSafe['type'];
+  @tracked address!: MerchantSafe['address'];
+  @tracked createdAt!: MerchantSafe['createdAt'];
+  @tracked tokens!: MerchantSafe['tokens'];
+  @tracked owners!: MerchantSafe['owners'];
+  @tracked accumulatedSpendValue!: MerchantSafe['accumulatedSpendValue'];
+  @tracked merchant!: MerchantSafe['merchant'];
+  @tracked infoDID: MerchantSafe['infoDID'];
+
+  constructor(safe: MerchantSafe) {
+    super();
+    this.update(safe);
+  }
+
+  update(safe: MerchantSafe) {
+    this.address = safe.address;
+    this.createdAt = safe.createdAt;
+    this.tokens = safe.tokens;
+    this.owners = safe.owners;
+    this.accumulatedSpendValue = safe.accumulatedSpendValue;
+    this.merchant = safe.merchant;
+    this.infoDID = safe.infoDID;
+  }
+}
+
+export class TrackedDepotSafe
+  extends TrackedSafe
+  implements RequiredKeepUndefined<DepotSafe>
+{
+  type = 'depot' as DepotSafe['type'];
+  @tracked address!: DepotSafe['address'];
```

Comment:
They could! But I didn't think there was much benefit to doing that, it generates Typescript and `update` method boilerplate code that we otherwise don't need. Are the property declarations too noisy?

Reply:
Refactored.